static/src/javascripts/projects/common/modules/navigation/newSticky.js

```diff
@@ -28,68 +28,112 @@ define([
+                var newAdHeightPromise = topAdRenderedPromise.then(getLatestAdHeight);
+
+                var getCachedAdHeight = function () {
+                    return Promise.race([newAdHeightPromise, oldAdHeightPromise]);
+                };
+
+                var render = function (state) {
+                    return fastdom.read(function () {
+                        var scrollY = window.scrollY;
+                        var scrollX = window.scrollX;
+
+                        return fastdom.write(function () {
+                            // Reset so we have a clean slate
+                            $header.css({
+                                'transition': '',
+                                'margin-top': ''
+                            });
+                            $adBanner.css({
+                                'position': '',
+                                'top': '',
+                                'max-height': ''
                             });
+
+                            // Set
+                            $header.css({ 'margin-top': state.adHeight });
+
+                            $adBanner.css({ 'max-height': state.adHeight });
+                            var userHasScrolledPastHeader = scrollY > headerHeight;
+                            if (userHasScrolledPastHeader) {
+                                $adBanner.css({
+                                    'position': 'absolute',
+                                    'top': headerHeight
+                                });
+                            } else {
+                                $adBanner.css({ 'position': 'fixed' });
+                            }
+
+                            var scrollIsAtTop = scrollY === 0;
+                            var diff = state.adHeight - state.previousAdHeight;
+                            var adHeightHasIncreased = diff > 0;
+                            if (!scrollIsAtTop && adHeightHasIncreased) {
+                                // If the user is not at the top and the ad height has increased,
+                                // we want to offset their scroll position
+                                window.scrollTo(scrollX, scrollY + diff);
+                            } else if (!state.firstRender) {
+                                // Otherwise we want to transition the change when it happens.
+                                // Avoid an initial transition when we apply the margin top for the first time
+                                $header.css('transition', 'margin-top 1s cubic-bezier(0, 0, 0, 0.985)');
+                            }
                         });
                     });
                 };
 
+                var update = (function () {
+                    var previousAdHeight;
+                    return function (options) {
+                        return getCachedAdHeight().then(function (adHeight) {
+                            return render({
```

Comment:
does render not return an instance of `fastdom`? is it thenable? i'm guessing it must be...?